src/control/server/storage/bdev/class.go

```diff
@@ -162,35 +162,45 @@ func prepBdevFile(l logging.Logger, c storage.BdevConfig) error {
 
 // genFromNvme takes NVMe device PCI addresses and generates config content
 // (output as string) from template.
-func genFromTempl(cfg storage.BdevConfig, templ string) (out bytes.Buffer, err error) {
+func genFromTempl(cfg *storage.BdevConfig, templ string) (out bytes.Buffer, err error) {
 	t := template.Must(template.New(confOut).Parse(templ))
 	err = t.Execute(&out, cfg)
+
 	return
 }
 
 // ClassProvider implements functionality for a given bdev class
 type ClassProvider struct {
 	log     logging.Logger
-	cfg     storage.BdevConfig
+	cfg     *storage.BdevConfig
 	cfgPath string
 	bdev    bdev
 }
 
+// NewClassProvider returns a new ClassProvider reference for given bdev type.
 func NewClassProvider(log logging.Logger, cfgDir string, cfg *storage.BdevConfig) (*ClassProvider, error) {
 	p := &ClassProvider{
 		log: log,
-		cfg: *cfg,
+		cfg: cfg,
 	}
 
 	switch cfg.Class {
-	case storage.BdevClassNone, storage.BdevClassNvme:
-		p.bdev = bdev{nvmeTempl, "", isEmptyList, isValidList, nilPrep}
+	case storage.BdevClassNone:
+		p.bdev = bdev{nvmeTempl, "", isEmptyList, isValidList, nilInit}
+	case storage.BdevClassNvme:
+		p.bdev = bdev{nvmeTempl, "NVME", isEmptyList, isValidList, nilInit}
+		if cfg.VmdEnabled {
```

Comment:
Yeah, so this class provider does introduce some awkwardness and I remember now why information needs to be conveyed between the backend and the class provider.

I do think it was a mistake in hindsight to expose the class provider outside of the bdev package -- its responsibilities should probably be integrated into bdevProvider and then the various methods would accept a per-ioserver storage config. This would allow us to completely compartmentalize details about bdev management to the bdev package.

In any case, fixing that is out of scope for this PR. I think at this point I'm not going to push for further improvements to the vmd device handling, with the understanding that some of this will be reworked in order to clean up the relationship between `bdev.spdkBackend` and `bdev.ClassProvider`.